Lesson 5-7: Simple character device

Derivability opened the pull request:
Implement character device driver with sysf and procfs interfaces.

Commit:
Lesson 5-7 part 1: Initial commit

Created empty main.c and Makefile.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 1: Add makefile content

Added makefile content for building target "character_device".

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 1: Implement buffer increase

Implemented buffer size increasing using module parameter.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 1: Implement buffer operations

Implemented buffer create and cleanup functions.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 1: Implement /dev operations

Implemented dev_read, dev_write, dev_open and dev_release functions.
Added variables "is_open", "used_memory".

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 1: Add init and exit functions

Added ChD_init and ChD_exit functions. Added "major" variable. Added
"GPL" license.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 2: Implement proc file operations

Implemented create_proc, cleanup_proc and proc_read functions.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 2: Add sysfs operations

Added sysfs_store function. Implemented creation sysfs directory in
ChD_init function.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 2: Fix memory allocation

Replaced kmalloc with kzalloc function. Reworked memory cleanup through
sysf.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 2: Fix memory allocation

Replaced kmalloc with kzalloc function. Reworked memory cleanup through
sysf.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 3: Create test script file

Created empty ChD_Test.sh file.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 3: Add varibles

Added variables for module name and path to module. Added variables for
procfs, sysf and /dev directory files path.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 3: Implement testing operations

Implemented reading/writing to device buffer, checking memory status
through procfs, cleaning buffer through sysfs and increasing buffer size
using module parameter.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 2: Fix tabulation

Fixed tabulation issues.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7 part 2: Add module information

Added module author, descriprion and version information.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7: Fix Makefile codestyle

Added empty line at the end of Makefile.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

galkinletter reviewed (changes requested):
Need fix

galkinletter commented:
02-KernelModules_CharacterDevice/main.c   ???

Need name for file  source  of driver  close to goal of driver

Commit:
Lesson 5-7: Change source filename

Change source file name. Edited makefile.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

galkinletter commented on 02-KernelModules_CharacterDevice/CharDevSource.c:
```diff
@@ -0,0 +1,269 @@
+
+//===init/exit functions===
+
+static int __init ChD_init(void){
```
```suggestion
static int __init ChD_init(void)
{
```
Please fix  Linux Kernel Coding Style
However, there is one special case, namely functions: they have the opening brace at the 
beginning of the next line, thus:
int function(int x)
{
        body of function
}

galkinletter reviewed (changes requested):
Need fix Linux Kernel Coding Style

Commit:
Lesson 5-7: Implement error handling

Implemented error handling int inti function.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

Commit:
Lesson 5-7: Add testing instructions

Added TestingScenario.md file. Removed hardcoded directory from
char_device_test.sh file.

Signed-off-by: Bogdan Shchogolev <bogdan993000@gmail.com>

galkinletter commented:
Ок
